### PR TITLE
load data: physical mode part1

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: ref #40499

Problem Summary:

### What is changed and how it works?
- ingegrate lightning physical mode into load data

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: executor/importer/chunk_process.go [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+	parser      mydump.Parser
+	chunkInfo   *checkpoints.ChunkCheckpoint
+	logger      *zap.Logger
+	kvsCh       chan []deliveredRow
+	dataWriter  *backend.LocalEngineWriter
+	indexWriter *backend.LocalEngineWriter
+
+	checksum verify.KVChecksum
+	encoder  kvEncoder
+	kvStore  tidbkv.Storage
```
Comment (GMHDBJD) [inline executor/importer/chunk_process.go:124]: can we use codec instead of kvstore？

Comment (D3Hunter) [reply executor/importer/chunk_process.go:124]: will use it to alloc auto-id in `common.AllocGlobalAutoID`

File: executor/importer/table_import.go [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+		StoreWriteBWLimit:       0,
+		ShouldCheckWriteStall:   false,
+		MaxOpenFiles:            int(util.GenRLimit()),
+		KeyspaceName:            keySpaceName,
+	}
+
+	tableMeta := &mydump.MDTableMeta{
+		DB:        e.DBName,
+		Name:      e.Table.Meta().Name.O,
+		DataFiles: e.toMyDumpFiles(),
```
Comment (GMHDBJD) [inline executor/importer/table_import.go:131]: Add TODO for `IsRowOrdered`?

Comment (D3Hunter) [reply executor/importer/table_import.go:131]: https://github.com/pingcap/tidb/pull/42817/commits/6ff452769d65ea6b584edb2e1294fbd14e05862e

**Review by GMHDBJD — APPROVED** [review]
LGTM

File: executor/importer/chunk_process.go [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+	b.indexChecksum = verify.NewKVChecksumWithKeyspace(b.codec)
+}
+
+func (b *deliverKVBatch) size() uint64 {
+	return b.dataChecksum.SumSize() + b.indexChecksum.SumSize()
+}
+
+func (b *deliverKVBatch) add(kvs *kv.Pairs) {
+	for _, pair := range kvs.Pairs {
+		if pair.Key[tablecodec.TableSplitKeyLen+1] == 'r' {
```
Comment (lance6716) [inline executor/importer/chunk_process.go:86]: Maybe we can use `tablecodec.IsRecordKey`

File: executor/importer/chunk_process.go [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+	checksum verify.KVChecksum
+	encoder  kvEncoder
+	kvCodec  tikv.Codec
+}
+
+func (p *chunkProcessor) process(ctx context.Context) error {
+	deliverCompleteCh := make(chan deliverResult)
+	go func() {
+		defer close(deliverCompleteCh)
+		err := p.deliverLoop(ctx)
```
Comment (lance6716) [inline executor/importer/chunk_process.go:129]: No need to change, but I prefer the error group pattern. When the job is finished, close the channel to notify the consumer. When any loop meets error, error group will automatically cancel the derived context and the error can be found by eg.Wait()

**Review by lance6716 — COMMENTED** [review]
5 / 11 files viewed

File: executor/importer/chunk_process.go (hunk 4 above)
Comment (D3Hunter) [reply executor/importer/chunk_process.go:129]: simplified from lightning code, will add a comment and do it later

File: executor/importer/chunk_process.go (hunk 3 above)
Comment (D3Hunter) [reply executor/importer/chunk_process.go]: [fix comments](https://github.com/pingcap/tidb/pull/42817/commits/c0dbaee38127111703e5fdd5ef1b3693ee1cf6be)

**Comment by D3Hunter** [comment]
multiple-valued index related test will add later

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: tests/realtikvtest/loaddatatest/load_data_test.go [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+	"github.com/pingcap/tidb/testkit"
+)
+
+func (s *mockGCSSuite) prepareAndUseDB(db string) {
+	s.tk.MustExec("drop database if exists " + db)
+	s.tk.MustExec("create database " + db)
+	s.tk.MustExec("use " + db)
+}
+
+func (s *mockGCSSuite) TestPhysicalMode() {
```
[the remark this answers is not in the document]
Comment (D3Hunter) [reply tests/realtikvtest/loaddatatest/load_data_test.go:31]: physical related can only run in `realtikvtest`(`check_dev_2`), we can move logical test here.

File: executor/importer/table_import.go [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+	"github.com/pingcap/tidb/util"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+)
+
+func prepareSortDir(e *LoadDataController) (string, error) {
+	tidbCfg := tidb.GetGlobalConfig()
+	// todo: add job id too
+	sortPathSuffix := "import-" + strconv.Itoa(int(tidbCfg.Port))
+	sortPath := filepath.Join(tidbCfg.TempDir, sortPathSuffix)
```
[the remark this answers is not in the document]
Comment (D3Hunter) [reply executor/importer/table_import.go:51]: copied from here

https://github.com/pingcap/tidb/blob/7c2c9922d100e3689c600ba281db6d3ee48d602b/ddl/ingest/env.go#L88

> will lightning's builtin disk quota conflicts with TiDB

yes, we should precheck whether's ongoing add-index.

File: executor/importer/table_import.go [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+	}
+
+	backendConfig := local.BackendConfig{
+		PDAddr:                  tidbCfg.Path,
+		LocalStoreDir:           dir,
+		MaxConnPerStore:         config.DefaultRangeConcurrency,
+		ConnCompressType:        config.CompressionNone,
+		WorkerConcurrency:       config.DefaultRangeConcurrency * 2,
+		KVWriteBatchSize:        config.KVWriteBatchSize,
+		CheckpointEnabled:       true,
```
[the remark this answers is not in the document]
Comment (D3Hunter) [reply executor/importer/table_import.go:116]: local backend report error when the sort-dir already exists & checkpoint disabled

will add a todo and fix it later

File: executor/importer/table_import.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (lance6716) [reply executor/importer/table_import.go:51]: No a big problem, we can check it later @tangenta PTAL at your convenience

**Review by lance6716 — APPROVED** [review]

**Pull request merged**